### PR TITLE
[tests] Refactor cache test to gTest

### PR DESCRIPTION
**Pull request by xinlipn**
Refactored cache test with GoogleTest framework

### CODE REVIEW

File: test/gtest/CMakeLists.txt
Comment (bghimireamd) [inline test/gtest/CMakeLists.txt:6]: Do we need these comments in future?

File: test/gtest/cache.cpp
Comment (bghimireamd) [inline test/gtest/cache.cpp:32]: For some reason I cannot see this file in https://github.com/ROCmSoftwarePlatform/MIOpen/tree/sl/google_test/test/gtest

File: test/gtest/CMakeLists.txt
Comment (bghimireamd) [inline test/gtest/CMakeLists.txt:3]: We could move `include(GoogleTest)` inside cmake/googletest.cmake?

File: Dockerfile
Comment (bghimireamd) [inline Dockerfile:21]: Just a nick pick. Why the space :)

File: test/gtest/cache.cpp
Comment (xinlipn) [reply test/gtest/cache.cpp:32]: Hey @bghimireamd , thank you for the comments. this header file is actually in an upper level :)
https://github.com/ROCmSoftwarePlatform/MIOpen/tree/sl/google_test/test

Comment (bghimireamd) [reply test/gtest/cache.cpp:32]: Oh! sorry I meant to say I cannot see file `test/gtest/cache.cpp` in https://github.com/ROCmSoftwarePlatform/MIOpen/tree/sl/google_test/test/gtest

File: Dockerfile
Comment (xinlipn) [reply Dockerfile:21]: Will remove the space :)

File: test/gtest/cache.cpp
Comment (xinlipn) [reply test/gtest/cache.cpp:32]: I see. I have checked out a fresh copy and switch to sl/test_cache branch. Seems it's there.
$ git clone https://github.com/ROCmSoftwarePlatform/MIOpen.git
$git switch sl/test_cache
$ls
CMakeLists.txt  cache.cpp  hello_test.cpp

Thanks

File: test/gtest/CMakeLists.txt
Comment (xinlipn) [reply test/gtest/CMakeLists.txt:3]: include(googletest) and include(GoogleTest) do look a bit of confusing. Moving to googletest.cmake indeed makes the code cleaner IMO. 

One thing is the following line in this CmakeLists.txt file is dependent on include(GoogleTest). Not sure if moving include(GoogleTest) to a different file makes dependency/future reading complicated. 

ln16:  gtest_discover_tests(test_${TEST_NAME})

@JehandadKhan could you also provide insights? Thanks

**Comment by xinlipn** [comment]
The CI failed at test_bn_3d_spatial_test, which seems to be related to PR 1424

https://github.com/ROCmSoftwarePlatform/MIOpen/issues/1424

File: test/gtest/CMakeLists.txt
Comment (JehandadKhan) [reply test/gtest/CMakeLists.txt:3]: I think I agree with Bibek, we can make it cleaner by moving the additional lines

File: test/gtest/CMakeLists.txt
Comment (JehandadKhan) [inline test/gtest/CMakeLists.txt:5]: Do we need rocblas for testing ? This has already been discovered in a higher level directory

**Comment by junliume** [comment]
@xinlipn the second to last CI passed but last commit has some CI issues:
http://micimaster.amd.com/blue/organizations/jenkins/MLLibs%2FMIOpen/detail/sl%2Ftest_cache/7/pipeline

**Comment by xinlipn** [comment]
> @xinlipn the second to last CI passed but last commit has some CI issues: http://micimaster.amd.com/blue/organizations/jenkins/MLLibs%2FMIOpen/detail/sl%2Ftest_cache/7/pipeline

Hi @junliume , a new commit should have fixed the link issue. Another CI is ongoing now.

http://micimaster.amd.com/blue/organizations/jenkins/MLLibs%2FMIOpen/detail/sl%2Ftest_cache/11/pipeline/

File: test/gtest/CMakeLists.txt
Comment (xinlipn) [reply test/gtest/CMakeLists.txt:5]: @JehandadKhan, find_package(rocblas)  is needed e.g. Target "test_api_convbiasactiv" links to target "roc::rocblas" 

> Do we need rocblas for testing ? This has already been discovered in a higher level directory

File: cmake/googletest.cmake
Comment (shurale-nkn) [inline cmake/googletest.cmake:27]: Why should we use this function? I don't want to update docker and cmake version just to make a git clone

**Review by shurale-nkn — CHANGES_REQUESTED** [review]

**Comment by junliume** [comment]
@xinlipn could you follow up with the review opinions above?

File: cmake/googletest.cmake
Comment (xinlipn) [reply cmake/googletest.cmake:27]: @shurale-nkn , thanks for the comments. If you refer to FetchContent_Declare, this is to declare a dependency on GoogleTest which will be downloaded from GitHub. GoogleTest is a newly integrated framework that's needed for unit testing. 

Only CMake 3.11 and later supports FetchContent. Does this make sense? Thanks

**Comment by xinlipn** [comment]
> @xinlipn could you follow up with the review opinions above?

@junliume Thanks for the reminder. Updated with a reply

File: cmake/googletest.cmake
Comment (shurale-nkn) [reply cmake/googletest.cmake:27]: @xinlipn 
This does not explain why it cannot be done without this function. And without new cmake.
What does this function give besides the implementation of git clone?
Our dockers already contain git and we used other repositories before GoogleTest .
What's the problem with using GoogleTest without this function?

Comment (xinlipn) [reply cmake/googletest.cmake:27]: @shurale-nkn 
FetchContent module makes content available immediately, allowing the configure step to use the content in commands like add_subdirectory(), include(), or file() operations. As part of FetchContent, FetchContent_Declare() is not just git clone, it records the options that describe how to populate the specified content. This feature is only supported by CMake 3.11 or later

Without which, CMake would throw errors below:

CMake Error at /usr/share/cmake-3.23/Modules/FetchContent.cmake:754 (message):
  No content details recorded for googletest
Call Stack (most recent call first):
  /usr/share/cmake-3.23/Modules/FetchContent.cmake:1140 (__FetchContent_getSavedDetails)
  /usr/share/cmake-3.23/Modules/FetchContent.cmake:1260 (FetchContent_Populate)
  cmake/googletest.cmake:35 (FetchContent_MakeAvailable)
  test/gtest/CMakeLists.txt:1 (include)

Comment (shurale-nkn) [reply cmake/googletest.cmake:27]: not sure why it's better than git submodule.

But ok, as it turned out, we are already using CMake 3.15, so your changes in Dockerfile are redundant.
If the PR will be merged earlier than the fix for issue #1706, no changes are required.

Comment (shurale-nkn) [reply cmake/googletest.cmake:27]: @xinlipn
wait, what is written in the error logs?
used a newer version of Cmake`/usr/share/cmake-3.23` and it returned an error at your function? 😄

Comment (xinlipn) [reply cmake/googletest.cmake:27]: Thanks @shurale-nkn , let me try on the new CMake version, in the meantime, do you have the logs with the erro?

Comment (shurale-nkn) [reply cmake/googletest.cmake:27]: @xinlipn 
I'm talking about your log from a previous comment

> @shurale-nkn FetchContent module makes content available immediately, allowing the configure step to use the content in commands like add_subdirectory(), include(), or file() operations. As part of FetchContent, FetchContent_Declare() is not just git clone, it records the options that describe how to populate the specified content, i.e. googletest in this case. This feature, i.e. FetechContent is only supported by CMake 3.11 or later
> 
> Without which, CMake would throw errors below:
> 
> CMake Error at /usr/share/cmake-3.23/Modules/FetchContent.cmake:754 (message): No content details recorded for googletest Call Stack (most recent call first): /usr/share/cmake-3.23/Modules/FetchContent.cmake:1140 (__FetchContent_getSavedDetails) /usr/share/cmake-3.23/Modules/FetchContent.cmake:1260 (FetchContent_Populate) cmake/googletest.cmake:35 (FetchContent_MakeAvailable) test/gtest/CMakeLists.txt:1 (include)
> 
> Not sure if this helps. Thanks

Comment (xinlipn) [reply cmake/googletest.cmake:27]: > @xinlipn wait, what is written in the error logs? used a newer version of Cmake`/usr/share/cmake-3.23` and it returned an error at your function? 😄

@shurale-nkn , I see. This CMake error was introduced after removing the FetchContent to address your comments earlier "What's the problem with using GoogleTest without this function?"

There's no CMake error with FetchContent :)

Comment (shurale-nkn) [reply cmake/googletest.cmake:27]: oh..., well, then it seems obvious that the second function with same prefix will not work and it will be necessary to assemble the package differently. no one canceled the build setup.

Comment (xinlipn) [reply cmake/googletest.cmake:27]: > oh..., well, then it seems obvious that the second function with same prefix will not work and it will be necessary to assemble the package differently. no one canceled the build setup.

Sorry if I wasn't clear.  The CMake error listed earlier was caused by removing FetchContent_Declare(). Not sure I totally get what you mean by "no one canceled the build setup". Could you elaborate?

File: test/gtest/CMakeLists.txt
Comment (shurale-nkn) [reply test/gtest/CMakeLists.txt:3]: @JehandadKhan @xinlipn 
related to this condition rocblas was not imported because it was OpenCL runtime stage.
line from 241 to 280

https://github.com/ROCmSoftwarePlatform/MIOpen/blob/4381ab1ef07b47416b2b9df14e47f6ddf57672e6/CMakeLists.txt#L241

https://github.com/ROCmSoftwarePlatform/MIOpen/blob/4381ab1ef07b47416b2b9df14e47f6ddf57672e6/CMakeLists.txt#L270-L275

The question arises why this test does not handle the flag MIOPEN_USE_ROCBLAS and MIOPEN_BACKEND.
https://github.com/ROCmSoftwarePlatform/MIOpen/blob/4381ab1ef07b47416b2b9df14e47f6ddf57672e6/src/CMakeLists.txt#L676

And I don't see where this test uses rocblas.

File: cmake/googletest.cmake
Comment (shurale-nkn) [reply cmake/googletest.cmake:27]: I meant that if we replaced the method of receiving the package, we need to replace the function responsible for its assembly. And for the assembly, it may be necessary to set the correct settings of the environment variables. 
But in this case, you will be sure of what you are doing and how the program should behave.

The "FetchContent" is not the only solution for dependency control.

**Review by bghimireamd — APPROVED** [review]

**Comment by junliume** [comment]
@xinlipn could you resolve the merge conflict?

File: Dockerfile
Comment (junliume) [inline Dockerfile:34]: We should remove these two lines.

File: cmake/googletest.cmake
Comment (junliume) [inline cmake/googletest.cmake:2]: Should it be here?

**Comment by junliume** [comment]
@xinlipn please check CI for DB verification and FIN stage

**Comment by xinlipn** [comment]
> @xinlipn please check CI for DB verification and FIN stage

@junliume renamed hello_test.cpp to hello_gtest.cpp to avoid conflicts with fin

File: test/gtest/cache.cpp
Comment (atamazov) [inline test/gtest/cache.cpp]: Please remove these comments.

> Please remove these comments.
Done

File: test/gtest/CMakeLists.txt
Comment (atamazov) [inline test/gtest/CMakeLists.txt]: This is not related to this PR.

**Review by atamazov — CHANGES_REQUESTED** [review]

File: test/gtest/CMakeLists.txt
Comment (xinlipn) [reply test/gtest/CMakeLists.txt]: Thanks for the comments. This is related to newly added tests since this branch is a bit dated.

File: Dockerfile
Comment (xinlipn) [reply Dockerfile]: Done

File: test/gtest/CMakeLists.txt
Comment (atamazov) [reply test/gtest/CMakeLists.txt]: I do not understand how dumpTensorTest is related to cache test. Please clearly indicate this change in the PR title and description or move it into separate PR. Thanks!

File: test/gtest/cache.cpp
Comment (atamazov) [reply test/gtest/cache.cpp]: Resolved.

File: test/gtest/CMakeLists.txt
Comment (xinlipn) [reply test/gtest/CMakeLists.txt]: Understood your concern. dumpTensorTest has been merged and this branch was dated. Trying to merge this branch back to develop and thus dumpTensorTest is pulled locally.

File: test/gtest/cache.cpp
Comment (bghimireamd) [inline test/gtest/cache.cpp:98]: Database name `""` (empty file name) are allowed? (Just curious, no code change needed)

**Review by bghimireamd — DISMISSED** [review]

File: test/gtest/CMakeLists.txt
Comment (shurale-nkn) [reply test/gtest/CMakeLists.txt]: and? when this will be fixed?

File: cmake/googletest.cmake
Comment (shurale-nkn) [inline cmake/googletest.cmake]: ?

Comment (shurale-nkn) [reply cmake/googletest.cmake]: Why we need this here

**Review by shurale-nkn — CHANGES_REQUESTED** [review]
FetchContent:

https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1652#discussion_r939050014

https://github.com/ROCmSoftwarePlatform/MIOpen/issues/1783#issuecomment-1257372834
https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1789#issuecomment-1258840710
https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1611/files#r912193677

**Review by atamazov — CHANGES_REQUESTED** [review]
This PR removes some other files, but it shouldn't

image

**Comment by xinlipn** [comment]
> This PR removes some other files, but it shouldn't
> 
> image

Thanks Artem, this is to address your previous comments about the unrelated changes to this PR in file "test/gtest/CMakeLists.txt" as below.   Can you clarify if these 5 files along with the changes below in test/gtest/CMakeLists.txt should be kept?

set(SOURCES ${SOURCES}
    "${CMAKE_CURRENT_SOURCE_DIR}/log_test_helper.cpp"
    )

if(MIOPEN_BACKEND_OPENCL)
  set(SKIP_TESTS dumpTensorTest)
endif()

**Comment by xinlipn** [comment]
> 1789 (comment)

Need a separate PR for FetchContent?

> FetchContent:
> 
> [#1652 (comment)](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1652#discussion_r939050014)
> 
> [#1783 (comment)](https://github.com/ROCmSoftwarePlatform/MIOpen/issues/1783#issuecomment-1257372834) [#1789 (comment)](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1789#issuecomment-1258840710) https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1611/files#r912193677

Changes in 1783 had been included in this PR

**Comment by atamazov** [comment]
@xinlipn The files you don't want to change should not be even listed in the changelist (but right now the PR removes them). Please revert these deletions from your branch. Perhaps it is worth to generally improve your git knowledge.

**Comment by xinlipn** [comment]
> 

The files were there because this branch is behind the develop.  The previous changes have been reverted. Deleted files are back now

File: cmake/googletest.cmake
Comment (JehandadKhan) [reply cmake/googletest.cmake]: @shurale-nkn  Thanks for the comment, let me fix this via requirements.txt in a followup PR, tracked in issue #1838

File: test/gtest/CMakeLists.txt
Comment (JehandadKhan) [reply test/gtest/CMakeLists.txt]: @xinlipn Please help us understand why this needed to be added to this PR?

**Review by JehandadKhan — CHANGES_REQUESTED** [review]
One minor issue, the rest looks fine.

File: test/gtest/CMakeLists.txt
Comment (xinlipn) [reply test/gtest/CMakeLists.txt]: image

@JehandadKhan will remove duplication. But if I delete dumpTensorTest and log_test_helper.cpp related changes all together. There will be make errors since this branch is dated and new files e.g. log_test_helper.cpp and dumpTensorTest.cpp etc.  had been merged to this branch

**Comment by xinlipn** [comment]
> One minor issue, the rest looks fine.

@JehandadKhan Thanks for the review. Those code had been removed

**Comment by xinlipn** [comment]
> 
@JehandadKhan these code in test/gtest/CMakeLists.txt had been removed

**Comment by xinlipn** [comment]
> FetchContent:
> 
> [#1652 (comment)](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1652#discussion_r939050014)
> 
> [#1783 (comment)](https://github.com/ROCmSoftwarePlatform/MIOpen/issues/1783#issuecomment-1257372834) [#1789 (comment)](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1789#issuecomment-1258840710) https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1611/files#r912193677

@shurale-nkn it appears this is handled in a different PR here 
 https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1789

**Review by JehandadKhan — DISMISSED** [review]

**Comment by JehandadKhan** [comment]
@junliume  What is blocking this PR ?

**Comment by atamazov** [comment]
> @junliume What is blocking this PR ?

https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1652#pullrequestreview-1148381067 by @shurale-nkn

**Comment by atamazov** [comment]
@JehandadKhan @xinlipn @junliume Please merge from current develop. Right now the baseline of this PR is too outdated (3 weeks). This may lead to build errors in develop after merging, e.g. "bugprone-implicit-widening-of-multiplication-result" because #1853 is not currently included into baseline.

**Comment by shurale-nkn** [comment]
My review expects that "fetchContent" will be replaced with a method corresponding to this repository.
This feature was added in a series of PR related to GoogleTest, which means it needs to be fixed in this series. IMHO

https://github.com/ROCmSoftwarePlatform/MIOpen/issues/1783#issuecomment-1257372834
"So I guess this is a quick fix, but the FetchContent with pulling googletest should be removed, so we are not downloading and building googletest everytime we configure cmake(ie every jenkins build/stage regardless of docker caching)."  (pfultz2)

**Comment by atamazov** [comment]
@junliume @JehandadKhan @xinlipn ⚠️ Quiet change. The refactored test does not obey SKIP_ALL_EXCEPT_TESTS anymore. As a result, the test will be executed when MIOpen is built with MIOPEN_TEST_INT8, MIOPEN_TEST_BFLOAT16 or MIOPEN_NO_GPU. The means that we are wasting time on our CI.

Albeit this is not important w.r.t. test_cache (because it is really quick), we shall take this into account when refactoring other tests.

**Comment by atamazov** [comment]
@shurale-nkn Please clearly indicate if this specific PR requires changes. If not, but GoogleTest in general needs some updates/fixes, then please open a ticket (like #1843) AND approve this PR or dismiss your review. Thanks!

**Comment by shurale-nkn** [comment]
> @shurale-nkn Please clearly indicate if this specific PR requires changes. If not, but GoogleTest in general needs some updates/fixes, then please open a ticket (like #1843) AND approve this PR or dismiss your review. Thanks!

The content of this PR has changed several times. At some point it also included the addition of 'fetchContent' function

**Comment by atamazov** [comment]
@shurale-nkn 
> The content of this PR has changed several times. At some point it also included the addition of 'fetchContent' function

It is not here anymore. Actually arrived to develop from #1726. Can you please open the ticket (plus give justification and set appropriate urgency)?

**Review by JehandadKhan — DISMISSED** [review]
Please remove the old test and merge develop into this branch

**Comment by xinlipn** [comment]
> Please remove the old test and merge develop into this branch

@JehandadKhan , this test seems to be disabled already in test/CMakeLists.txt as below

https://github.com/ROCmSoftwarePlatform/MIOpen/blob/312bc9d39aaab143526889e53ae956cc52d2efb3/test/CMakeLists.txt#L457

**Comment by JehandadKhan** [comment]
@xinlipn Please merge develop in this branch, so that the updated CI can run.

**Review by junliume — APPROVED** [review]

**Comment by junliume** [comment]
@shurale-nkn @JehandadKhan ping for last round of review before merging this one.

**Review by shurale-nkn — APPROVED** [review]
In the current state, this is only renaming functions from c++ to gtest format.
No objections.

**Pull request merged**
